src/proxy.c

```diff
@@ -264,88 +280,81 @@ cc_proxy_msg_read (GIOChannel *source, GIOCondition condition,
 	}
 
 	if (bytes_read == 0) {
-		g_critical("failed to read lenght");
+		g_critical("failed to read length");
 		return false;
 	}
 
-	if (proxy_data->oob_fd) {
-		/* check if oob data was received */
+	total_bytes_read += bytes_read;
+
+	if (proxy_data->oob_fd && !proxy_data->oob_fd_found) {
+		/*
+		 * We expect OOB data that we have not found yet.
+		 * The OOB message must come first, as a 1 byte
+		 * only message.
+		 */
 		cmsg = CMSG_FIRSTHDR(&msg);
-		if (cmsg) {
-			data = CMSG_DATA(cmsg);
-			if (data) {
-				*proxy_data->oob_fd = *((int*) data);
-				g_message("oob fd read from proxy socket %d",
-					*proxy_data->oob_fd);
-				fd_read = true;
-			}
+		if (!cmsg) {
+			g_critical("could not read the control message");
+			goto out;
 		}
-	}
 
-	/* only out-of-band data was received */
-	if (bytes_read < sizeof (buf_len) || iov_buffer[0] == 'F' ) {
-		goto out1;
+		data = CMSG_DATA(cmsg);
+		if (! (data && iov_buffer[0] == 'F')) {
```

Comment:
This "magic `F`" is important. But it's also mentioned explicitly in various places (4 times). From a maintenance perspective, I'd be inclined to create a `#define` for it and change all hard-coded references to use the define (including all the references in comments). This would bring `proxy.c` sort of in line with `proxy.go` which defines `fileFlag` (although `proxy.go` could do a bit better in this respect too as there are various mentions of "`F`" scattered about in comments).